File: discr_segm.py
"""
Module that discretizes a fracture in to small sub fractures(nodes)
Rewritten from a matlab script by Stephan de Hoop

Author: Mieke Daemen/Jieran Li
Date: 22/11/2020
"""

import math as m
import numpy as np


def discr_segm(x_coor, y_coor, lencal):
    """
    This function takes a the x and y -coordinates of a fractures and returns a matrix with sub fractures.
    :param x_coor: an array with two elements that contains the x coordinates of the fracture.
                        the first element, x_coor[0], is the x-coordinate of one vertex
                        the second element, x_coor[1], is the x-coordinate of the other vertex
    :param y-coor: an array with two elements that contains the y coordinates of the fracture.
                        the first element, x_coor[0], is the x-coordinate of one vertex
                        the second element, x_coor[1], is the x-coordinate of the other vertex
    :param lencal: the length of the sub segments.
    :return: seg_mat: a matrix with the coordinates of the individual sub segments.

    """
    # length of a fracture
    len_frac = m.sqrt((x_coor[1]-x_coor[0])**2 + (y_coor[1]-y_coor[0])**2)
    # number of fractures in one fracture, round and make at least equal to 1
    num_segm = int(round(len_frac / lencal))
    if num_segm == 0:
        num_segm = 1

    seg_mat = np.zeros((num_segm, 4))
    for i in range(num_segm):
        seg_mat[i,0] = x_coor[0]+(i/num_segm)*(x_coor[1]-x_coor[0])
        seg_mat[i,1] = y_coor[0]+(i/num_segm)*(y_coor[1]-y_coor[0])
        seg_mat[i,2] = x_coor[0]+((i+1)/num_segm)*(x_coor[1]-x_coor[0])
        seg_mat[i,3] = y_coor[0]+((i+1)/num_segm)*(y_coor[1]-y_coor[0])
    return seg_mat

File: discr_segm.m
function [segm_mat] = discr_segm(x_coor, y_coor, char_len)

% Calculate length of big fracture-segment:
len_frac = sqrt( (x_coor(2) - x_coor(1)).^2 + ...
                 (y_coor(2) - y_coor(1)).^2 );

% Find number of smaller segments (round and make at least 1 segm!):
num_segm = round(len_frac/char_len);
if num_segm == 0
    % Very small segment, will get merged anyway!
    num_segm = 1;
end

% Recalculate length of smaller segments (based on floored #segm):
len_segm = len_frac / num_segm;

% Find nodes-list of each smaller segment:
% [x_1, y_1, x_2, y_2]
segm_mat = zeros(num_segm, 4);

for ith_dis_segm = 1:num_segm
    % Add each small segment to list:
    % First segment, fix coordinate to original coordinate:
    segm_mat(ith_dis_segm, 1:2) = [x_coor(1)+((ith_dis_segm-1)/num_segm)*(x_coor(2)-x_coor(1)), ...
                                   y_coor(1)+((ith_dis_segm-1)/num_segm)*(y_coor(2)-y_coor(1))];
    segm_mat(ith_dis_segm, 3:4) = [x_coor(1)+(ith_dis_segm/num_segm)*(x_coor(2)-x_coor(1)), ...
                                   y_coor(1)+(ith_dis_segm/num_segm)*(y_coor(2)-y_coor(1))];
end

if num_segm == 0
    % Store normal segment:
    segm_mat = [x_coor(1), y_coor(1), x_coor(2), y_coor(2)];
end